node_modules/oae-principals/lib/libraries/members.js

```diff
@@ -0,0 +1,293 @@
+            return log().error({'err': err, 'userId': user.id}, 'An error occurred while trying to update all members libraries to which a user belongs');
+        }
+
+        var entries = _.map(allRoles, function(groupRole) {
+            return {
+                'id': groupRole.id,
+                'oldRank': _getMembersLibraryRank(groupRole.id, oldPrincipal, groupRole.role),
+                'newRank': _getMembersLibraryRank(groupRole.id, updatedPrincipal, groupRole.role),
+                'resource': updatedPrincipal,
+                'value': groupRole.role
+            };
+        });
+
+        return LibraryAPI.Index.update(PrincipalsConstants.library.MEMBERS_INDEX_NAME, entries);
+    });
+};
+
+/*!
+ * When a user or group updates their profile or visibility, we have to update their rank in the
+ * members libraries to which they belong
+ */
+PrincipalsEmitter.on(PrincipalsConstants.events.UPDATED_USER, _updatePrincipalInMembersLibraries);
+PrincipalsEmitter.on(PrincipalsConstants.events.UPDATED_GROUP, _updatePrincipalInMembersLibraries);
+
+/*!
+ * When a group's members are updated, we should update the members library
+ */
+PrincipalsEmitter.when(PrincipalsConstants.events.UPDATED_GROUP_MEMBERS, function(ctx, group, memberChanges, newMemberIds, updatedMemberIds, removedMemberIds, callback) {
+    // When group members are updated (users are removed / roles updated, etc...) it's just easier
+    // to purge the library unless there is good reason to do surgical library updates (e.g., known
+    // performance issues with many group members updates)
+    LibraryAPI.Index.purge(PrincipalsConstants.library.MEMBERS_INDEX_NAME, group.id, function(err) {
+        if (err) {
+            log().error({'err': err, 'groupId': group.id}, 'An unexpected error occurred trying to purge a group members library');
+            return callback(err);
+        }
+
+        return callback();
+    });
+});
+
+/*!
+ * When someone joins a group, we should add them to the group's members library
+ */
+PrincipalsEmitter.when(PrincipalsConstants.events.JOINED_GROUP, function(ctx, group, joinRole, callback) {
+    var entry = {
+        'id': group.id,
+        'rank': _getMembersLibraryRank(group.id, ctx.user(), joinRole),
+        'resource': ctx.user(),
+        'value': joinRole
+    };
+
+    LibraryAPI.Index.insert(PrincipalsConstants.library.MEMBERS_INDEX_NAME, [entry], function(err) {
+        if (err) {
+            log().error({'err': err, 'groupId': group.id, 'userId': ctx.user().id}, 'An unexpected error occurred trying to insert a user into a group members library');
+            return callback(err);
+        }
+
+        return callback();
+    });
+});
+
+/*!
+ * When someone leaves a group, we should remove them from the group's members library. We don't
+ * have to do this with a synchronous `when` handler because the user generally won't notice within
+ * such small latency if the members library hasn't been updated
```

Comment:
I don't think so. When joining a group, I would think looking at the members list is something they may do next. If they don't see themselves there, then they could think something is wrong.

Contrarily, if a user leaves a group, I don't imagine they'd care to go look at the members library afterward.

That's a lot of hand-waving, though, but if anyone is really passionate about it I could switch it to be synchronous.